src/core/pointcloud/qgspointcloudeditingindex.cpp

```diff
@@ -114,15 +124,63 @@ QgsPointCloudBlockRequest *QgsPointCloudEditingIndex::asyncNodeData( const QgsPo
   return nullptr;
 }
 
-bool QgsPointCloudEditingIndex::commitChanges()
+bool QgsPointCloudEditingIndex::commitChanges( QString *errorMessage )
 {
   if ( !isModified() )
     return true;
 
-  if ( !mIndex.updateNodeData( mEditedNodeData ) )
+  QHash<QgsPointCloudNodeId, QgsCopcUpdate::UpdatedChunk> updatedChunks;
+  for ( auto it = mEditedNodeData.constBegin(); it != mEditedNodeData.constEnd(); ++it )
+  {
+    QgsPointCloudNodeId n = it.key();
+    // right now we're assuming there's no change of point count
+    qint32 nodePointCount = static_cast<qint32>( getNode( n ).pointCount() );
+    updatedChunks[n] = QgsCopcUpdate::UpdatedChunk{ nodePointCount, it.value() };
+  }
+
+  QFileInfo fileInfo( mUri );
+  QString outputFilename = fileInfo.dir().filePath( fileInfo.baseName() + "-update.copc.laz" );
+
+  if ( !QgsCopcUpdate::writeUpdatedFile( mUri, outputFilename, updatedChunks, errorMessage ) )
+  {
     return false;
+  }
+
+  // reset the underlying index - we will reload it at the end
+  QgsCopcPointCloudIndex *copcIndex = static_cast<QgsCopcPointCloudIndex *>( mIndex.get() );
+  copcIndex->reset();
+
+  QString originalFilename = fileInfo.dir().filePath( fileInfo.baseName() + "-original.copc.laz" );
+  if ( !QFile::rename( mUri, originalFilename ) )
+  {
+    if ( errorMessage )
+      *errorMessage = "Rename of the old COPC failed!";
+    QFile::remove( outputFilename );
+    return false;
+  }
+
+  if ( !QFile::rename( outputFilename, mUri ) )
+  {
+    if ( errorMessage )
+      *errorMessage = "Rename of the new COPC failed!";
+    QFile::rename( originalFilename, mUri );
+    QFile::remove( outputFilename );
+    return false;
+  }
+
+  if ( !QFile::remove( originalFilename ) )
+  {
+    if ( errorMessage )
+      *errorMessage = "Removal of the old COPC failed!";
+    // TODO: cleanup here as well?
```

Comment:
Is the comment still relevant?

Reply:
yeah, it's unclear what to do... removal of the old (now renamed) file failed, but in fact from the user's point of view the commit succeeded - so I'm not sure what the most appropriate here. Probably it does not matter because this would never happen, and if there's a problem, it would fail earlier :slightly_smiling_face: